### ISSUE TITLE
Clarification for gradient memory in mixed precision training

### ISSUE THREAD

**Issue by SumanthRH** [stas00/ml-engineering#23]
Hi @stas00 ,

I was going over your excellent notes on [model memory usage](https://github.com/stas00/ml-engineering/blob/master/performance/software.md#anatomy-of-models-memory-usage) and I noticed the following:

> Gradients
> 
> 4 bytes * number of parameters for either fp32 or mixed precision training (gradients are almost always kept in fp32).
> 2 bytes * number of parameters for more recent works where half-precision is used

I was trying to see how you got the 4 bytes number for mixed precision! From what I've seen in the [mixed precision](https://arxiv.org/abs/1710.03740) paper and the [ZeRO](https://arxiv.org/abs/1910.02054) paper, the gradients are kept in half-precision and should take 2 bytes / parameter. However, during the optimization step, the gradients are converted to FP32 - but I assumed this happens on the fly and thus you don't have to write the full FP32 gradient tensor into HBM. 

When I was benchmarking memory usage using some of your [older BERT-based benchmarking code](https://huggingface.co/docs/transformers/v4.20.1/perf_train_gpu_one), the memory usage with FP16 training works out to be 11GB on my machine. If I use the 4 bytes number for gradients instead of 2 bytes, the estimated memory will end up coming closer to the 11GB amount! I was initially assuming this discrepancy is some overhead with temporary buffers (instead of number of bytes for gradient elements to be 4). I'm just curious to know more about this 4 byte number, and even better if you're aware of the implementation details for this in Apex or DeepSpeed. Thank you!

**Comment by stas00** [comment]
There are a few places where the Deepspeed implementation of ZeRO upcasts grads to fp32, 

1. it does it before stepping as it tries not to meddle with the optim API, so any optimizer can be used. It'd have to meddle with the optim API to perform the upcasting on the fly.
2. If you use `communication_data_type=fp32`, to avoid the lossy reduction of grads, it'll also upcast to fp32, though only doing it to the amount of params fitting into the current comms bucket (so typically far less than all params). 

So there are at least those 2 cases where 4 bytes are used. This might change in the future.

Now those upcasts can really be considered temp memory usage as they use peak memory, but practically this is 4 bytes per param.

I don't know how Apex does it.

How do you measure memory usage, are you taking into the account the peak memory? Which would have led to OOM if there weren't enough spare.

You can try this package in a jupyter notebook https://github.com/stas00/ipyexperiments if you want automated reports of peak and normal memory usages. The package is very trivial to setup and you will just need to load your code into a jupyter notebook, which sometimes can be more difficult to setup. But you could use even something as simple as https://github.com/Syllo/nvtop to get the peak memory usage.

**Comment by SumanthRH** [comment]
Got it! I understand now. Part of the issue was that I did not understand `torch.autocast` properly. After inspecting training step with mixed precision, it's clear that the full FP32 gradients are stored in HBM and this matches the memory consumption seen as well. 

> If you use communication_data_type=fp32, to avoid the lossy reduction of grads, it'll also upcast to fp32, though only doing it to the amount of params fitting into the current reduction bucket (so typically far less than all params).

Ah interesting. I believe I'll probably need to dig into the code to actually understand this fully :) 

> How do you measure memory usage, are you taking into the account the peak memory? Which would have led to OOM if there weren't enough spare.

Hmm I was using `pynvml` to measure current memory usage. You're right that peak memory is not accounted for here. `ipyexperiments` looks great! I'll check it out!

Thank you so much for the detailed reply!

**Comment by stas00** [comment]
There are 2 ways that I know of to measure peak GPU memory:
- using `pynvml` to take frequent snapshots (imprecise, but good enough) - that's what `ipyexperiments` does
- https://pytorch.org/docs/stable/generated/torch.cuda.max_memory_allocated.html - (but there is only one counter, so if more than application uses it and resets the counter - the results are unpredictable.)
